src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,428 @@
+				+ SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES
+				== IETF_CHACHA_NONCESIZE
+		);
+		let mut state = Self {
+			key: subkey_and_nonce(&key, &nonce).0,
+			counter: 1,
+			inonce: [0u8; SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES],
+		};
+		state
+			.inonce
+			.copy_from_slice(&nonce.as_ref()[HCHACHA_NONCESIZE..]);
+		state
+	}
+
+	/// prepare a new push or pull operation
+	pub fn init(&mut self, key: chacha20Key, nonce: Nonce) {
+		*self = SecretStreamXChaCha20Poly1305::new(key, nonce);
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// derives a new secret key used for encryption/decryption
+	pub fn rekey(&mut self) -> Result<(), UnknownCryptoError> {
+		let mut new_key_and_inonce =
+			[0u8; CHACHA_KEYSIZE + SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES];
+
+		new_key_and_inonce[..CHACHA_KEYSIZE].copy_from_slice(self.key.unprotected_as_bytes());
+		new_key_and_inonce[CHACHA_KEYSIZE..].copy_from_slice(&self.inonce);
+
+		chacha20_enc_in_place(&self.key, &self.get_nonce(), 0, &mut new_key_and_inonce)?;
+
+		self.key = chacha20Key::from_slice(&new_key_and_inonce[..CHACHA_KEYSIZE]).unwrap();
+		self.inonce
+			.copy_from_slice(&new_key_and_inonce[CHACHA_KEYSIZE..]);
+
+		self.counter = 1;
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// //Ecryptes a message
+	pub fn encrypt_message(
+		&mut self,
+		plaintext: &[u8],
+		ad: &[u8],
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		let msglen = plaintext.len();
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		let mut block = [0u8; BLOCKSIZE];
+		let adlen = ad.len();
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+
+		block[0] = tag.bits();
+		chacha20_enc_in_place(&self.key, &self.get_nonce(), 1, &mut block)?;
+		dst_out[0] = block[0];
+
+		chacha20_enc(
+			&self.key,
+			&self.get_nonce(),
+			2,
+			plaintext,
+			&mut dst_out[cipherpos..],
+		)?;
+
+		let mac = self
+			.generate_auth_tag(dst_out, ad, msglen, &block, adlen, cipherpos)
+			.unwrap();
+		debug_assert!(dst_out.len() >= macpos + mac.get_length());
+		dst_out[macpos..(macpos + mac.get_length())].copy_from_slice(mac.unprotected_as_bytes());
+		xor_buf8(self.inonce.as_mut(), &dst_out[macpos..macpos + 8]);
+		self.counter = self.counter.wrapping_add(1);
+		if bool::from(tag.bits().ct_eq(&Tag::REKEY.bits()) | self.counter.ct_eq(&0u32)) {
+			self.rekey()?;
+		}
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// Decryptes a message
+	pub fn decrypt_message(
+		&mut self,
+		cipher: &[u8],
+		ad: &[u8],
+		plaintext_out: &mut [u8],
+		tag_out: &mut Tag,
+	) -> Result<(), UnknownCryptoError> {
+		let msglen = cipher.len() - SECRETSTREAM_XCHACHA20POLY1305_ABYTES;
```

Comment:
The length check must come before this subtraction, otherwise it will silently overflow in release builds.